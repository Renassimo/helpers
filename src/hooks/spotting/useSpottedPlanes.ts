import { useContext } from 'react';

import SpottingContext from '@/contexts/spotting';

const useSpottedPlanes = () => useContext(SpottingContext);

export default useSpottedPlanes;
