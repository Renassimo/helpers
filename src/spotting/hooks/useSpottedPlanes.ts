import { useContext } from 'react';

import SpottingContext from '@/spotting/contexts/SpottingContext';

const useSpottedPlanes = () => useContext(SpottingContext);

export default useSpottedPlanes;
