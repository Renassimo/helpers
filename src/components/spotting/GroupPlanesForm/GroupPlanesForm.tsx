import React, { ChangeEvent, useMemo } from 'react';

import useSpottedPlanes from '@/hooks/spotting/useSpottedPlanes';

import { useTheme } from '@mui/material/styles';

import TextField from '@mui/material/TextField';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';

const HASHTAG_LIMIT = 30;

const GroupPlanesForm = () => {
  const {
    groupDescription,
    groupHashtags,
    groupName,
    setGroupDescription,
    setGroupHashtags,
    setGroupName,
  } = useSpottedPlanes();

  const descriptionInputName = `group-description`;
  const hashtagsInputName = `group-hashtags`;
  const groupNameInputName = `group-name`;

  const onDescriptionChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setGroupDescription(event.target.value);
  };
  const onHashtagsChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setGroupHashtags(event.target.value);
  };
  const onGroupNameChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setGroupName(event.target.value);
  };

  const theme = useTheme();
  const descriptionHashtagsCount = useMemo(
    () => groupDescription.split('#').length - 1,
    [groupDescription]
  );
  const hashtagsCount = useMemo(
    () => groupHashtags.split('#').length - 1,
    [groupHashtags]
  );
  const totalHashTags = useMemo(
    () => descriptionHashtagsCount + hashtagsCount,
    [descriptionHashtagsCount, hashtagsCount]
  );
  const hashtagsExceeded = useMemo(
    () => totalHashTags > HASHTAG_LIMIT,
    [totalHashTags]
  );
  const hashtagColor = useMemo(
    () =>
      hashtagsExceeded
        ? theme.palette.warning.main
        : theme.palette.primary.main,
    [hashtagsExceeded, theme.palette.primary]
  );

  return (
    <>
      <TextField
        id={descriptionInputName}
        name={descriptionInputName}
        value={groupDescription}
        onChange={onDescriptionChange}
        multiline
        fullWidth
        margin="dense"
        variant="standard"
        placeholder="Description"
      />
      <Typography variant="caption" color={hashtagColor}>
        Hashtags: {descriptionHashtagsCount}
      </Typography>
      <TextField
        id={hashtagsInputName}
        name={hashtagsInputName}
        value={groupHashtags}
        onChange={onHashtagsChange}
        multiline
        fullWidth
        margin="dense"
        variant="standard"
        placeholder="Hashtags"
      />
      <Typography variant="caption" color={hashtagColor}>
        Hashtags: {hashtagsCount}
      </Typography>
      <Box display="flex" justifyContent="flex-end">
        <TextField
          id={groupNameInputName}
          name={groupNameInputName}
          value={groupName}
          onChange={onGroupNameChange}
          margin="dense"
          variant="standard"
          placeholder="Group name"
        />
      </Box>
    </>
  );
};

export default GroupPlanesForm;
