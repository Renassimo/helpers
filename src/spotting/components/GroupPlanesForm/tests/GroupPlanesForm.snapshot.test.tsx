import renderWithTheme from '@/tests/helpers';

import useSpottedPlanes from '@/spotting/hooks/useSpottedPlanes';

import GroupPlanesForm from '@/spotting/components/GroupPlanesForm';

jest.mock('@/spotting/hooks/useSpottedPlanes');

describe('GroupPlanesForm', () => {
  const mockedGroupDescription = 'Group description #descriptionHashtag';
  let mockedGroupHashtags = '';
  const mockedGroupName = 'Group name';
  const mockedSetGroupDescription = jest.fn();
  const mockedSetGroupHashtags = jest.fn();
  const mockedSetGroupName = jest.fn();

  beforeEach(() => {
    (useSpottedPlanes as unknown as jest.Mock).mockImplementation(() => ({
      groupDescription: mockedGroupDescription,
      groupHashtags: mockedGroupHashtags,
      groupName: mockedGroupName,
      setGroupDescription: mockedSetGroupDescription,
      setGroupHashtags: mockedSetGroupHashtags,
      setGroupName: mockedSetGroupName,
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should rendered successfully', () => {
    // Arrange
    mockedGroupHashtags = '#GroupHashtags';
    // Act
    const { container } = renderWithTheme(<GroupPlanesForm />);
    // Assert
    expect(container).toMatchSnapshot();
  });

  describe('When exceeded hashtags', () => {
    test('should rendered successfully', () => {
      // Arrange
      mockedGroupHashtags = `#GroupHashtags ${'#hashtag'.repeat(29)}`;
      // Act
      const { container } = renderWithTheme(<GroupPlanesForm />);
      // Assert
      expect(container).toMatchSnapshot();
    });
  });
});
