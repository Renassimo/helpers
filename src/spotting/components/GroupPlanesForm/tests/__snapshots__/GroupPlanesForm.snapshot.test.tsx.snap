// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupPlanesForm When exceeded hashtags should rendered successfully 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-10lvcmn-MuiInputBase-root-MuiInput-root"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        id="group-description"
        name="group-description"
        placeholder="Description"
        style="height: 0px; overflow: hidden;"
      >
        Group description #descriptionHashtag
      </textarea>
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-caption css-1nr2uw6-MuiTypography-root"
  >
    Hashtags: 
    1
  </span>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-10lvcmn-MuiInputBase-root-MuiInput-root"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        id="group-hashtags"
        name="group-hashtags"
        placeholder="Hashtags"
        style="height: 0px; overflow: hidden;"
      >
        #GroupHashtags #hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag#hashtag
      </textarea>
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-caption css-1nr2uw6-MuiTypography-root"
  >
    Hashtags: 
    30
  </span>
  <div
    class="MuiBox-root css-1bvc4cc"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root css-4mzek5-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1eed5fa-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="group-name"
          name="group-name"
          placeholder="Group name"
          type="text"
          value="Group name"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`GroupPlanesForm should rendered successfully 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-10lvcmn-MuiInputBase-root-MuiInput-root"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        id="group-description"
        name="group-description"
        placeholder="Description"
        style="height: 0px; overflow: hidden;"
      >
        Group description #descriptionHashtag
      </textarea>
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Hashtags: 
    1
  </span>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-10lvcmn-MuiInputBase-root-MuiInput-root"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        id="group-hashtags"
        name="group-hashtags"
        placeholder="Hashtags"
        style="height: 0px; overflow: hidden;"
      >
        #GroupHashtags
      </textarea>
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-66dh3a-MuiInputBase-input-MuiInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Hashtags: 
    1
  </span>
  <div
    class="MuiBox-root css-1bvc4cc"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root css-4mzek5-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1eed5fa-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="group-name"
          name="group-name"
          placeholder="Group name"
          type="text"
          value="Group name"
        />
      </div>
    </div>
  </div>
</div>
`;
